### PR TITLE
add credential pre-flight check against proof requests

### PR DESCRIPTION
**Pull request by chipp**
## Why
Pre-flight check to determine whether the credential store can satisfy a proof request before attempting proof generation. Also validates the structural integrity of the constraint expression, mirroring `ProofRequest::validate_response`.

## What
- New `credential_check` module with `check_proof_request` (uniffi-exported)
- `CredentialCheckError` with `Storage`, `ConstraintTooDeep`, and `ConstraintTooLarge` variants
- `CredentialValidation` and `CredentialItemValidation` records expose per-item and overall satisfaction status
- Evaluates `Any`/`All`/`Enumerate` constraint expressions against non-expired credentials in the store
- Validates constraint depth ≤ 2 and node count ≤ `MAX_CONSTRAINT_NODES` before evaluating
  - can be simplified when https://github.com/worldcoin/world-id-protocol/pull/732 is merged and released
  - and walletkit bumps world-id-core to 0.11.x
- Exposes `genesis_issued_at` from the DB in `CredentialRecord` and uses it together with `expires_at` to validate `RequestItem` time constraints, preventing false-positive `is_satisfied` results
- 18 unit tests covering all constraint variants, error cases, and edge cases

## Risk
Low — additive; the only changes to existing code are exposing `genesis_issued_at` in `CredentialRecord` (query + struct) and the `credential_check` module re-export.

### CODE REVIEW

File: walletkit-core/src/lib.rs [hunk 1]
```diff
@@ -131,6 +131,9 @@ pub use user_agent::{UserAgent, UserAgentBuilder};
 /// Proof requests and responses in World ID v4.
 pub mod requests;
 
+/// Pre-flight validation of whether stored credentials can satisfy a [`requests::ProofRequest`].
+pub mod credential_check;
```
Comment (sujoyg) [inline walletkit-core/src/lib.rs]: Nit: Consider making the name more descriptive, for e.g.  `proof_request_credential_constraints_check`?

File: walletkit-core/src/credential_check.rs [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+    #[error("constraint nesting exceeds maximum allowed depth")]
+    ConstraintTooDeep,
+    /// The constraint expression exceeds the maximum node count.
+    #[error("constraints exceed maximum allowed size")]
+    ConstraintTooLarge,
+}
+
+/// Satisfaction status for a single credential request item.
+#[derive(Debug, Clone, uniffi::Record)]
+pub struct CredentialItemValidation {
```
Comment (sujoyg) [inline walletkit-core/src/credential_check.rs]: Nit: Item or Constraint or CredentialStore?

File: walletkit-core/src/credential_check.rs [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+    pub has_credential: bool,
+}
+
+/// Result of validating whether a [`CredentialStore`] can satisfy a [`ProofRequest`].
+#[derive(Debug, Clone, uniffi::Record)]
+pub struct CredentialValidation {
+    /// `true` when the constraint tree (or all items, if no constraints) is satisfied.
+    pub is_satisfied: bool,
+    /// Per-request-item satisfaction detail.
+    pub items: Vec<CredentialItemValidation>,
```
Comment (sujoyg) [inline walletkit-core/src/credential_check.rs]: Nit: It is not clear which `items` will be included. For example, is it the minimal list of credential stores that satisfy the constraint? What happens if more than one credential satisfies the constraint?

Is it set when is_satisfied is false? If yes, what is included in items in that case?

It may be clearer to separate the field that returns the missing credentials from satisfying credentials. Further, if constraints are satisfied then there may not be any need to include items at all. 

Consider linking to a short design doc of various scenarios or adding it here as documentation.

File: walletkit-core/src/credential_check.rs [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+/// When constraints are present the expression is validated for depth/size
+/// (mirroring `ProofRequest::validate_response`) and then evaluated.
+///
+/// # Errors
+///
+/// - [`CredentialCheckError::Storage`] if the credential query fails.
+/// - [`CredentialCheckError::ConstraintTooDeep`] if the constraint tree exceeds depth 2.
+/// - [`CredentialCheckError::ConstraintTooLarge`] if the constraint tree exceeds the node limit.
+#[uniffi::export]
+pub fn check_proof_request(
```
Comment (sujoyg) [inline walletkit-core/src/credential_check.rs]: Nit: Consider making the name more descriptive. Example ProofRequestConstraintSatisfyingCredentialsCheck

File: walletkit-core/src/credential_check.rs [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+    request: &ProofRequest,
+    store: &CredentialStore,
+    now: u64,
+) -> Result<CredentialValidation, CredentialCheckError> {
```
Comment (sujoyg) [inline walletkit-core/src/credential_check.rs]: Nit: ProofRequestConstraintSatisfyingCredentials or something like to make it more descriptive?

File: walletkit-core/src/issuers/recovery_bindings_manager.rs [hunk 6]
```diff
@@ -248,6 +248,7 @@ impl RecoveryBindingManager {
 }
 
 #[cfg(test)]
+#[cfg(all(not(target_arch = "wasm32"), feature = "embed-zkeys"))]
```
Comment (Dzejkop) [inline walletkit-core/src/issuers/recovery_bindings_manager.rs:251]: /otto why is this necessary?

**Review by sujoyg — DISMISSED** [review]

**Comment by Dzejkop** [comment]
@chipp since https://github.com/worldcoin/world-id-protocol/pull/732 is already merged can we simplify

File: walletkit-core/src/proof_request_credential_constraints_check.rs#L139C1-C74

```rust
// TODO: replace with `request.0.validate_constraints()?` once
```

File: walletkit-core/src/proof_request_credential_constraints_check.rs [hunk 7]
```diff
@@ -0,0 +1,673 @@
...
+
+    let mut check_results: Vec<CredentialConstraintsCheckItem> = Vec::new();
+    for item in &request.0.requests {
+        let expires_min = item.expires_at_min.unwrap_or(request.0.created_at);
+        let genesis_min = item.genesis_issued_at_min.unwrap_or(0);
+
+        let has_credential =
+            by_schema.get(&item.issuer_schema_id).is_some_and(|creds| {
+                creds.iter().any(|r| {
+                    r.expires_at > expires_min && r.genesis_issued_at >= genesis_min
```
[the remark this answers is not in the document]
Comment (Dzejkop) [reply walletkit-core/src/proof_request_credential_constraints_check.rs:148]: it's fine

File: walletkit-core/src/proof_request_credential_constraints_check.rs [hunk 8]
```diff
@@ -0,0 +1,643 @@
...
+//!
+//! # Per-item evaluation
+//!
+//! For each [`RequestItem`] the check verifies that the store contains at least one
+//! credential that is:
+//!
+//! 1. **Not expired** — `expires_at > now`.
+//! 2. **Fresh enough** — `genesis_issued_at >= genesis_issued_at_min` (defaults to 0
+//!    when the request item omits the field, meaning any issuance time is accepted).
+//! 3. **Long-lived enough** — `expires_at >= expires_at_min` (defaults to
```
[the remark this answers is not in the document]
Comment (Dzejkop) [reply walletkit-core/src/proof_request_credential_constraints_check.rs]: @chipp minor, but looks valid

**Review by Dzejkop — APPROVED** [review]

File: walletkit-core/src/proof_request_credential_constraints_check.rs [hunk 9]
```diff
@@ -0,0 +1,696 @@
...
+/// See the [module-level documentation](self) for a full description of the
+/// evaluation logic and intended usage.
+///
+/// # Errors
+///
+/// - [`CredentialConstraintsCheckError::Storage`] if the credential store query fails.
+/// - [`CredentialConstraintsCheckError::ConstraintTooDeep`] if the constraint tree exceeds depth 2.
+/// - [`CredentialConstraintsCheckError::ConstraintTooLarge`] if the constraint tree exceeds the node limit.
+#[uniffi::export]
+pub fn check_credentials_against_proof_request(
```
Comment (paolodamico) [inline walletkit-core/src/proof_request_credential_constraints_check.rs:128]: we shouldn't be exposing this high level order function, we should be extending ProofRequest to have a `from_{x}` to be idiomatic with Rust.

File: walletkit-core/src/proof_request_credential_constraints_check.rs [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+//! iterate `check_results` and surface items where `has_credential` is `false` to
+//! tell the user which credentials are missing or do not meet the request's time
+//! constraints.
+//!
+//! # Examples
+//!
+//! The table below shows how constraints, available credentials, and per-item
+//! results combine. ✓ = `has_credential: true`, ✗ = `has_credential: false`.
+//!
+//! | Request items         | Constraints                      | Credentials in store  | `is_satisfied` | `check_results`                    |
```
Comment (paolodamico) [inline walletkit-core/src/proof_request_credential_constraints_check.rs:55]: Let's simplify the description and table here, there's a lot more information than needed and most of the information should come from https://docs.rs/world-id-primitives/latest/world_id_primitives/request/struct.ProofRequest.html. walletkit is just an SDK layer

File: walletkit-core/src/proof_request_credential_constraints_check.rs [hunk 11]
```diff
@@ -0,0 +1,696 @@
...
+    };
+
+    Ok(CredentialConstraintsCheckResult {
+        is_satisfied,
+        check_results,
+    })
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (paolodamico) [inline walletkit-core/src/proof_request_credential_constraints_check.rs:185]: most of these tests should be dropped. this behavior is already covered in the `world-id-primitives` crate

File: walletkit-core/src/proof_request_credential_constraints_check.rs (hunk 9 above)
Comment (chipp) [reply walletkit-core/src/proof_request_credential_constraints_check.rs:128]: addressed in #411
